lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php

```diff
@@ -0,0 +1,836 @@
+		$title = WP_Font_Family_Utils::get_font_face_slug( $settings );
+
+		$prepared_post->post_type    = $this->post_type;
+		$prepared_post->post_parent  = $request['font_family_id'];
+		$prepared_post->post_status  = 'publish';
+		$prepared_post->post_title   = $title;
+		$prepared_post->post_name    = sanitize_title( $title );
+		$prepared_post->post_content = wp_json_encode( $settings );
+
+		return $prepared_post;
+	}
+
+	/**
+	 * Handles the upload of a font file using wp_handle_upload().
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param array $file Single file item from $_FILES.
+	 * @return array Array containing uploaded file attributes on success, or error on failure.
+	 */
+	protected function handle_font_file_upload( $file ) {
+		add_filter( 'upload_mimes', array( 'WP_Font_Library', 'set_allowed_mime_types' ) );
+		add_filter( 'upload_dir', 'wp_get_font_dir' );
+
+		$overrides = array(
+			'upload_error_handler' => array( $this, 'handle_font_file_upload_error' ),
+			// Arbitrary string to avoid the is_uploaded_file() check applied
+			// when using 'wp_handle_upload'.
+			'action'               => 'wp_handle_font_upload',
+			// Not testing a form submission.
+			'test_form'            => false,
+			// Seems mime type for files that are not images cannot be tested.
+			// See wp_check_filetype_and_ext().
+			'test_type'            => true,
+			// Only allow uploading font files for this request.
+			'mimes'                => WP_Font_Library::get_expected_font_mime_types_per_php_version(),
+		);
+
+		$uploaded_file = wp_handle_upload( $file, $overrides );
+
+		remove_filter( 'upload_dir', 'wp_get_font_dir' );
+		remove_filter( 'upload_mimes', array( 'WP_Font_Library', 'set_allowed_mime_types' ) );
+
+		return $uploaded_file;
+	}
+
+	/**
+	 * Handles file upload error.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param array  $file    File upload data.
+	 * @param string $message Error message from wp_handle_upload().
+	 * @return WP_Error WP_Error object.
+	 */
+	public function handle_font_file_upload_error( $file, $message ) {
+		$status = 500;
+		$code   = 'rest_font_upload_unknown_error';
+
+		if ( __( 'Sorry, you are not allowed to upload this file type.', 'default' ) === $message ) {
+			$status = 400;
+			$code   = 'rest_font_upload_invalid_file_type';
+		}
+
+		return new WP_Error( $code, $message, array( 'status' => $status ) );
+	}
+
+	/**
+	* Returns relative path to an uploaded font file.
+	*
+	* The path is relative to the current fonts dir.
+	*
+	* @since 6.5.0
+	* @access private
+	*
+	* @param string $path Full path to the file.
+	* @return string Relative path on success, unchanged path on failure.
+	*/
+	protected function relative_fonts_path( $path ) {
+		$new_path = $path;
+
+		$fonts_dir = wp_get_font_dir();
+		if ( str_starts_with( $new_path, $fonts_dir['path'] ) ) {
+			$new_path = str_replace( $fonts_dir, '', $new_path );
+			$new_path = ltrim( $new_path, '/' );
+		}
+
+		return $new_path;
+	}
+
```

Comment:
I think this function should not be part of the rest controller class but part of WP_Font_Library or WP_Font_Family. It won't be used just by this controller but for the WP_REST_Font_Families_Controller too to write the preview assets.